controllers/templates.go

```diff
@@ -60,34 +58,83 @@ func Builder() echo.HandlerFunc {
 		order := c.QueryParam("order")
 
 		objectId := primitive.NewObjectID()
-		component := components.Builder(inputType, objectId.Hex(), order)
+		component := components.TemplateBuilderInput(inputType, objectId.Hex(), order)
 
 		return renderNoBase(c, component)
 	}
 }
 
 func Save(database *mongo.Database, client *mongo.Client) echo.HandlerFunc {
 	return func(c echo.Context) error {
-		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
-		defer cancel()
-
-		if err := c.Request().ParseForm(); err != nil {
-			return err
-		}
-
-		formValues, err := c.FormParams()
+		formValues, err := formparser.ValidateFormValues(c)
 		if err != nil {
 			return err
 		}
 
-		db.SaveTemplate(database, ctx, formparser.TemplateFromForm(formValues))
+		// TODO: Add validations for template builder
+		db.SaveTemplate(database, models.Template{}.FromForm(formValues))
 
-		return renderNoBase(c, components.Save())
+		return renderNoBase(c, components.Save("Template"))
 	}
 }
 
-func Delete() echo.HandlerFunc {
+func DeleteInput() echo.HandlerFunc {
 	return func(c echo.Context) error {
-		return renderNoBase(c, components.Delete())
+		return renderNoBase(c, components.DeleteInput())
+	}
+}
+
+func CompleteTemplate(database *mongo.Database) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		id := c.Param("id")
+
+		template := db.GetTemplate(database, id)
+
+		questionViews := formparser.QuestionsToView(template.Questions)
+
+		return renderNoBase(c, components.Complete(template, questionViews))
+	}
+}
+
+func Complete(database *mongo.Database) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		id := c.Param("id")
+
+		template := db.GetTemplate(database, id)
+
+		questionViews := formparser.QuestionsToView(template.Questions)
+
+		formValues, err := formparser.ValidateFormValues(c)
+		if err != nil {
+			return err
+		}
+
+		questionViews = formparser.ApplyValidations(questionViews, formValues)
+
+		if formparser.HasErrors(questionViews) {
+			return renderNoBase(c, components.Complete(template, questionViews))
+		}
+
+		// TODO: Extract all this
+		cookie, err := c.Cookie("token")
+		if err != nil {
+			views.AddFlash(c, "You must be logged in to access that page", views.FlashError)
+
+			return redirect(c, "/login")
+		}
+
+		tokenString := cookie.Value
+
+		parsedToken, err := auth.ParseToken(tokenString)
+
+		if err != nil {
+			views.AddFlash(c, "Invalid or expired token", views.FlashError)
+
+			return redirect(c, "/login")
+		}
+
+		db.SaveAnswer(database, models.Answer{}.FromForm(template.ID, parsedToken["sub"].(string), questionViews))
+
+		return renderNoBase(c, components.Save("Response to template"))
```

Comment:
How will this work if the page is reloaded? We might want to redirect or change the URL here

Reply:
Same with the template build creation

Reply:
https://github.com/hay-i/gooal/issues/76